builtin/credential/aws/path_login.go

```diff
@@ -128,13 +128,211 @@ needs to be supplied along with 'identity' parameter.`,
 			logical.AliasLookaheadOperation: &framework.PathOperation{
 				Callback: b.pathLoginUpdate,
 			},
+			logical.ResolveRoleOperation: &framework.PathOperation{
+				Callback: b.pathLoginResolveRole,
+			},
 		},
 
 		HelpSynopsis:    pathLoginSyn,
 		HelpDescription: pathLoginDesc,
 	}
 }
 
+func (b *backend) pathLoginResolveRole(ctx context.Context, req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	anyEc2, allEc2 := hasValuesForEc2Auth(data)
+	anyIam, allIam := hasValuesForIamAuth(data)
+	switch {
+	case anyEc2 && anyIam:
+		return logical.ErrorResponse("supplied auth values for both ec2 and iam auth types"), nil
+	case anyEc2 && !allEc2:
+		return logical.ErrorResponse("supplied some of the auth values for the ec2 auth type but not all"), nil
+	case anyEc2:
+		return b.pathLoginResolveRoleEc2(ctx, req, data)
+	case anyIam && !allIam:
+		return logical.ErrorResponse("supplied some of the auth values for the iam auth type but not all"), nil
+	case anyIam:
+		return b.pathLoginResolveRoleIam(ctx, req, data)
+	default:
+		return logical.ErrorResponse("didn't supply required authentication values"), nil
+	}
+}
+
+func (b *backend) pathLoginEc2GetRoleNameAndIdentityDocumentParsed(ctx context.Context, req *logical.Request, data *framework.FieldData) (string, *identityDocument, *logical.Response, error) {
```

Comment:
This is a nit so feel free to ignore :). The method name here is a bit long and might be hard to grok at quick glance. Maybe something like `pathLoginEc2GetRoleNameAndIdentityDoc` which is slightly shorter is still descriptive enough to get the intent across? What do you think?

Reply:
Cool, thanks!